homeassistant/helpers/dispatcher.py

```diff
@@ -26,6 +30,39 @@ def remove_dispatcher() -> None:
     return remove_dispatcher
 
 
+def wrap_callback(func: Callable[..., Any]) -> Callable[[], None]:
+    """Decorate a signal callback to catch and log exceptions."""
+    def log_exception(signal: str, *args: Any) -> None:
+        module_name = inspect.getmodule(inspect.trace()[1][0]).__name__
+        # Do not print the wrapper in the traceback
+        frames = len(inspect.trace()) - 1
+        err = traceback.format_exc(-frames)
+        logging.getLogger(module_name).error(
+            "Exception in %s when dispatching '%s': %s\n%s",
+            func.__name__, signal, args, err)
+
+    wrapper_func = None
```

Comment:
Can you make a function to generate this instead of us having to end up copy pasting it everywhere.

```python
catch_callback(callback, log_exception)
```

I wonder if we even can have `log_exception` be generated and we just pass in a lambda to generate the string to print when erroring.

```python
catch_callback(callback, lambda args, err: "dispatching '{}': {}".format(signal, args))
```

Reply:
Sure, where does that function go? In utils somewhere?

Reply:
I added the common stuff to util/logging.py
(This PR now depends on #19977)